[EP-2362] Save contact info after sign up

canac opened the pull request:
# Changes

* Replace the `onStateChange` binding with dedicated `onSignUp` and `onSignIn` bindings
* After sign up, use the order service to save the donor details to the user's account. If saving fails, show the contact info form with the fields from sign up pre-populated.

I made this a PR instead of committing directly so it's easier for you to see what changed and see if I'm going in the right direction with this.

dr-bizz commented on src/common/components/contactInfoModal/contactInfoModal.tpl.html:
```diff
@@ -1,4 +1,4 @@
-<contact-info submitted="$ctrl.submitted" on-submit="$ctrl.onSubmit(success)"></contact-info>
+<contact-info submitted="$ctrl.submitted" on-submit="$ctrl.onSubmit(success)" donor-details="$ctrl.signUpDonorDetails"></contact-info>
```
Where is `signUpDonorDetails` defined? Were you also meant to commit changes to the contactInfoModal.component.js

canac replied:
It is set in `checkDonorDetails` if there was an error saving the donor details

dr-bizz commented on src/common/components/registerAccountModal/registerAccountModal.tpl.html:
```diff
@@ -31,9 +31,8 @@ <h3 translate>Create Your Account</h3>
-    on-state-change="$ctrl.stateChanged(state)"
-    on-success="$ctrl.onIdentitySuccess()"
-    on-failure="$ctrl.onIdentityFailure()"
+    on-sign-up="$ctrl.onSignUp(donorDetails)"
+    on-sign-in="$ctrl.onSignIn()"
```
You also need to update the file `src/common/services/session/sessionModal.tpl.html` as signUpModal is also used there

canac replied:
Good catch, that should be fixed now

dr-bizz commented on src/common/components/signUpModal/signUpModal.component.js:
```diff
@@ -148,13 +148,36 @@ class SignUpModalController {
           }
+        },
+        // TODO: make sure that this callback is only called after the sign up completes successfully
+        postSubmit: (postData, onSuccess) => {
```
It looks like this is fired after the Registration API is fired to Okta, but before registration is completed. I think that is okay, bt as you said, we will need to check

dr-bizz commented on src/common/components/signUpModal/signUpModal.component.js:
```diff
@@ -148,13 +148,36 @@ class SignUpModalController {
 
     this.oktaSignInWidget.on('afterRender', this.afterRenderFn.bind(this))
+
+    this.$scope.$apply(() => this.onSignUp())
```
Could you explain why you need to run `this.onSignUp()` here?

canac replied:
Good catch! That was accidentally left in for testing

dr-bizz reviewed (approved):
It looks great! I have a few questions, but I don't want to block you so I will approve the PR.

canac commented:
@dr-bizz Sorry for the force-push. I rebased on top of #1132, combined several fixup commits, and added the organization fields you added. I'll go through it myself more in tomorrow morning to see if I missed anything, but I'm pushing for you to take a look at too when you get a chance.

dr-bizz commented on src/common/components/registerAccountModal/registerAccountModal.component.js:
```diff
@@ -117,7 +134,29 @@ class RegisterAccountModalController {
+
+      // Pass donorDetails to the next step
+      return Observable.of(donorDetails)
+    }).subscribe({
```
I switched `do` to `catch` as it was easier to read, but these changes are just suggestions based on preference. 

You are correct, you don't need to write `Observable.throw(error)` if you write `do`

canac commented on src/common/components/registerAccountModal/registerAccountModal.component.spec.js:
```diff
@@ -169,16 +169,16 @@ describe('registerAccountModal', function () {
         it('changes state to \'contact-info\'', () => {
           $ctrl.orderService.getDonorDetails.mockImplementation(() => Observable.of({ 'registration-state': 'COMPLETED' }))
-          $ctrl.getDonorDetails()
+          $ctrl.checkDonorDetails()
```
Good idea. I added some tests.
